sematic/db/models/external_resource.py

```diff
@@ -194,6 +196,41 @@ def set_resource(self, resource: AbstractExternalResource) -> None:
 
     resource = property(get_resource, set_resource)
 
+    def force_to_terminal_state(self, reason: str):
```

Comment:
From what I can tell, this does not force the clean up of the external resource infrastructure-wise; it only forces the DB model of the resource to be clean. Is that correct?

Reply:
Correct. The idea is that you would normally not use "force" when cleaning. Then things will only enter a terminal state if the external resources are actually cleaned. However, there are cases where the metadata can get into a state where it can't be updated, even if the actual infrastructure has been cleaned. As an example, if you move the python path for a particular abstract external resource while it is orphaned, then the metadata can't be updated. We actually had several cases of this on dev1, because that's the DB I used during development (though it could be possible for users as well if they switch from EE to non-ee, we move a class, or they have their own implementations which they then move). When you get resources in such a state, you can manually clean the infrastructure and use "force" to clean up the metadata.